File: src/swift/4_BehavioralPatterns/9_State/Handmade.swift
import Foundation
import XCTest

enum State
{
  case offHook
  case connecting
  case connected
  case onHold
}

enum Trigger
{
  case callDialed
  case hungUp
  case callConnected
  case placedOnHold
  case takenOffHold
  case leftMessage
}

let rules = [
  State.offHook: [
    (Trigger.callDialed, State.connecting)
  ],
  State.connecting: [
    (Trigger.hungUp, State.offHook),
    (Trigger.callConnected, State.connected)
  ],
  State.connected: [
    (Trigger.leftMessage, State.offHook),
    (Trigger.hungUp, State.offHook),
    (Trigger.placedOnHold, State.onHold)
  ],
  State.onHold: [
    (Trigger.takenOffHold, State.connected),
    (Trigger.hungUp, State.offHook)
  ]
]

func main()
{
  var state = State.offHook // starting state
  while true
  {
    print("The phone is currently \(state)")
    print("Select a trigger:")

    for i in 0..<rules[state]!.count
    {
      let (t, _) = rules[state]![i]
      print("\(i). \(t)")
    }

    if let input = Int(readLine()!)
    {
      let (_, s) = rules[state]![input]
      state = s
    }
  }
}

main()

File: src/csharp/4_BehavioralPatterns/9_State/Handmade.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using static System.Console;

namespace DesignPatterns
{
  public enum State
  {
    OffHook,
    Connecting,
    Connected,
    OnHold
  }

  public enum Trigger
  {
    CallDialed,
    HungUp,
    CallConnected,
    PlacedOnHold,
    TakenOffHold,
    LeftMessage
  }

  class Demo
  {
    private static Dictionary<State, List<(Trigger, State)>> rules
      = new Dictionary<State, List<(Trigger, State)>>
      {
        [State.OffHook] = new List<(Trigger, State)>
        {
          (Trigger.CallDialed, State.Connecting)
        },
        [State.Connecting] = new List<(Trigger, State)>
        {
          (Trigger.HungUp, State.OffHook),
          (Trigger.CallConnected, State.Connected)
        },
        [State.Connected] = new List<(Trigger, State)>
        {
          (Trigger.LeftMessage, State.OffHook),
          (Trigger.HungUp, State.OffHook),
          (Trigger.PlacedOnHold, State.OnHold)
        },
        [State.OnHold] = new List<(Trigger, State)>
        {
          (Trigger.TakenOffHold, State.Connected),
          (Trigger.HungUp, State.OffHook)
        }
      };

    static void Main(string[] args)
    {
      var state = State.OffHook;
      while (true)
      {
        WriteLine($"The phone is currently {state}");
        WriteLine("Select a trigger:");

        // foreach to for
        for (var i = 0; i < rules[state].Count; i++)
        {
          var (t, _) = rules[state][i];
          WriteLine($"{i}. {t}");
        }


        int input = int.Parse(Console.ReadLine());

        var (_, s) = rules[state][input];
        state = s;
      }
    }
  }
}